jax/experimental/fused_attention_stableHLO.py

```diff
@@ -0,0 +1,561 @@
+    # Only support fp16 and bf16 here
+    assert query_dtype in [jnp.float16, jnp.bfloat16]
+    # Q, K and V must be 4-D tensors
+    assert len(query.shape) == len(key.shape) == len(value.shape) == 4
+ 
+    batch_size, q_seq_len, num_heads, head_dim = query.shape
+    _, kv_seq_len, _, _ = key.shape
+    output_shape = (batch_size, q_seq_len, num_heads, head_dim)
+    activation_shape = (batch_size, num_heads, q_seq_len, kv_seq_len)
+    return (
+        ShapedArray(output_shape, query_dtype),  # output
+        ShapedArray(activation_shape, query_dtype),  # activation
+    )
+
+def _dot_product_attention_bwd_abstract(query, key, value, bias, mask, activation, fwd_output, grad_output,
+    *, scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask):
+    query_dtype = dtypes.canonicalize_dtype(query.dtype)
+    key_dtype = dtypes.canonicalize_dtype(key.dtype)
+    value_dtype = dtypes.canonicalize_dtype(value.dtype)
+    # Q, K and V must have the same data type
+    assert query_dtype == key_dtype == value_dtype
+    # Only support fp16 and bf16 here
+    assert query_dtype in [jnp.float16, jnp.bfloat16]
+    # Q, K and V must be 4-D tensors
+    assert len(query.shape) == len(key.shape) == len(value.shape) == 4
+    return (
+        ShapedArray(
+            query.shape, query_dtype
+        ),  # grad query
+        ShapedArray(
+            key.shape, key_dtype
+        ),  # grad key
+        ShapedArray(
+            value.shape, value_dtype
+        ),  # part value
+    )
+
+def _dot_product_attention_fwd_cuda_lowering(ctx, query, key, value, bias, mask,
+    scale, seed, dropout_rate, variadic_args, is_flash_attention, is_causal_mask):
+    query_type = ir.RankedTensorType(query.type)
+    query_shape = query_type.shape
+    key_type = ir.RankedTensorType(key.type)
+    key_shape = key_type.shape
+    value_type = ir.RankedTensorType(value.type)
+    value_shape = value_type.shape
+    
+    batch_size, q_seq_len, num_heads, head_dim = query_shape
+    _, kv_seq_len, _, _ = key_shape
+
+    output_shape = (batch_size, num_heads, q_seq_len, head_dim)
+    output_layout = (3, 1, 2, 0)
+    output_transpose_perm = mlir.dense_int_elements((0, 2, 1, 3))
+    activation_shape = (batch_size, num_heads, q_seq_len, kv_seq_len)
+    scratch_shape = (0,)
+    scratch_type = ir.IntegerType.get_unsigned(8)
+    # get backend config
+    backend_config = create_dot_product_attention_backend_config(batch_size, num_heads, q_seq_len, kv_seq_len, query_type.element_type, scale, seed, dropout_rate, is_flash_attention, is_causal_mask, False)
+    # {Q, K, V, mask*, bias*}
+    # {output, scratch, activation*}
+    has_dropout = dropout_rate > 0
+    has_bias, has_mask = variadic_args
+    operands = [query, key, value]
+    if has_mask:
+        operands.append(mask)
+    if has_bias:
+        operands.append(bias)
+    # get custom call name
+    custom_call_name = get_custom_call_name(has_bias, has_mask, has_dropout, False)
```

Comment:
Should this return the expected number of input and you can assert this?
Just an extra check in case of future changes.